### PR TITLE
Import document even if ExportMarker is not first item

### LINKED ISSUE #64: Transfer document not recognized if there is a bookmark in the import placeholder
Report https://forums.zotero.org/discussion/98929/switch-document-processor-but-still-saying-it-was-created-with-a-different-document-processor

This is reliably reproducible.

### PR DESCRIPTION
**Pull request by raffaem**
Import document even if ExportMarker is not the first item, altough it
must still be the first text item. Fix #64.

### CODE REVIEW

**Comment by adomasven** [comment]
@raffaem Pinging you in case you missed my comments here somehow, since you were active on other branches yesterday.

**Comment by raffaem** [comment]
> @raffaem Pinging you in case you missed my comments here somehow, since you were active on other branches yesterday.

I don't see comments :( ?

**Review by adomasven — CHANGES_REQUESTED** [review]
Sorry, I didn't submit the review comments!

**Comment by raffaem** [comment]
Maybe I misunderstood how it works.

Take `checkforExportMarker`.

The variable `xParaEnum` is an iterator over paragraphs, right?

Then `xParaPortionAccess` is the first paragraph of the document?

Then `xPortionEnum` is an iterator over *elements* (text, bookmarks, tables, images, etc.) of the first paragraph?

**Comment by raffaem** [comment]
ok done. Here are the documents for testing.

[zotero_#61_documents.zip](https://github.com/zotero/zotero-libreoffice-integration/files/9315018/zotero_.61_documents.zip)

File: build/source/org/zotero/integration/ooo/comp/Document.java [hunk 1]
```diff
@@ -248,13 +248,21 @@ public boolean importDocument() throws Exception {
...
 			text.removeTextContent(paragraph);
 		}
+		// remove last paragraph
+		Object lastParagraphObj = null;
+		while(xParaEnum.hasMoreElements()) {
+			lastParagraphObj = xParaEnum.nextElement();
+		}
+		XTextContent lastParagraph = UnoRuntime.queryInterface(XTextContent.class, lastParagraphObj);
+		text.removeTextContent(lastParagraph);
+		// return
```
Comment (adomasven) [inline build/source/org/zotero/integration/ooo/comp/Document.java:265]: I know the import-export workflow inside LibreOffice creates a trailing paragraph, but this is not true for other word processors, and generally removing the final paragraph with other document processors can throw errors in certain cases, so the code just allows the trailing paragraph to stay as that's a lesser evil.

File: build/source/org/zotero/integration/ooo/comp/Document.java [hunk 2]
```diff
@@ -248,13 +248,21 @@ public boolean importDocument() throws Exception {
 				XEnumerationAccess.class, text);
 		XEnumeration xParaEnum = xParaAccess.createEnumeration();
 		ArrayList<XTextContent> removeParagraphs = new ArrayList<XTextContent>();
-		for (int i = 0 ; i < 3 && xParaEnum.hasMoreElements(); i++) {
+		for (int i = 0 ; i < 4 && xParaEnum.hasMoreElements(); i++) {
```
Comment (adomasven) [inline build/source/org/zotero/integration/ooo/comp/Document.java:251]: This line shouldn't change.

File: build/source/org/zotero/integration/ooo/comp/Document.java [hunk 3]
```diff
@@ -731,38 +739,48 @@ public XTextViewCursor getSelection() {
 	}
 	
 	private boolean checkForExportMarker() throws Exception {
-		XEnumerationAccess xParaAccess = UnoRuntime.queryInterface(
+		// get paragraphs
+		XEnumerationAccess paragraphs = UnoRuntime.queryInterface(
```
Comment (adomasven) [inline build/source/org/zotero/integration/ooo/comp/Document.java:743]: I know variable name is not that straightforward, but this is the naming convention for UNO objects across the codebase, which differentiates them from our own object types, so let's maintain the consistency.

**Review by adomasven — CHANGES_REQUESTED** [review]

File: build/source/org/zotero/integration/ooo/comp/Document.java (hunk 2 above)
Comment (raffaem) [reply build/source/org/zotero/integration/ooo/comp/Document.java:251]: But the export creates 4 new paragraphs, not 3.

Please see the test documents where the first line is marked "MY FIRST LINE".

This should be the first line after importing too!

Comment (adomasven) [reply build/source/org/zotero/integration/ooo/comp/Document.java]: Ok, I just checked the other plugins to make sure this won't cause issues, and yes, all others remove 4, so this is fine.

File: build/source/org/zotero/integration/ooo/comp/Document.java [hunk 4]
```diff
@@ -255,6 +255,7 @@ public boolean importDocument() throws Exception {
 		for (XTextContent paragraph : removeParagraphs) {
 			text.removeTextContent(paragraph);
 		}
+		// return
```
Comment (adomasven) [inline build/source/org/zotero/integration/ooo/comp/Document.java:258]: No need for the comment here. It adds no additional clarity.

File: build/source/org/zotero/integration/ooo/comp/Document.java (hunk 2 above)
Comment (raffaem) [reply build/source/org/zotero/integration/ooo/comp/Document.java]: you drive me crazy

Comment (adomasven) [reply build/source/org/zotero/integration/ooo/comp/Document.java:251]: Sorry, but the general rule is don't fix what's not broken. I spent a ton of time on making the import/export code cross-compatible between word processors, so any changes have to be warranted and require cross-testing and checking.

**Comment by adomasven** [comment]
Thanks for working through this! Sorry for being so anal about the review, but codebase consistency and clarity is something that makes long-term maintenance of the code, especially as it is handed off between developers easier, and software can die quickly if it becomes unmaintainable. We'll release your changes in the Zotero 6 beta soon.

**Comment by adomasven** [comment]
And good job on making this contribution. We don't often get such high-quality PRs, especially for document processor plugins, where they require specific weird domain knowledge.

File: build/source/org/zotero/integration/ooo/comp/Document.java (hunk 2 above)
Comment (raffaem) [reply build/source/org/zotero/integration/ooo/comp/Document.java:251]: > Sorry, but the general rule is don't fix what's not broken. I spent a ton of time on making the import/export code cross-compatible between word processors, so any changes have to be warranted and require cross-testing and checking.

I'm kidding! I know the difficulties in cross-compatibility! ;)

**Comment by raffaem** [comment]
> Thanks for working through this! Sorry for being so anal about the review, but codebase consistency and clarity is something that makes long-term maintenance of the code, especially as it is handed off between developers easier, and software can die quickly if it becomes unmaintainable. We'll release your changes in the Zotero 6 beta soon.

Thank you for the wonderful software 😁 

But what are the conventions for the variables that refer to UNO objects? They have to start with an `x`? Anything else?

**Comment by adomasven** [comment]
> But what are the conventions for the variables that refer to UNO objects? They have to start with an `x`? Anything else?

Idk if that's the convention that was recommended originally by OpenOffice/LibreOffice developers for plugin devs, and I inherited this codebase from another dev, and the convention was already in place, so we try to maintain it.

**Pull request merged**